application/detections/detection_interface.hpp

```diff
@@ -0,0 +1,18 @@
+#pragma once
+#include "detection_algorithms.hpp"
+#include <memory>
+#include <opencv2/core.hpp>
+#include <vector>
+
+namespace campo {
+namespace detections {
+
+class IDetectionAlgorithm {
+public:
+  virtual ~IDetectionAlgorithm() = default;
+  virtual void detect(const cv::Mat &image,
+                      std::vector<cv::Rect> &detectedObjects) = 0;
```

Comment:
If you want to have a very lean interface, use std::function. You have only one function - detect, so you can store it in a lambda. You can keep std::vector<std::function> to store many different detection algorithms.
And you don't need so many hpp files. One that contains a factory and all algorithms should be enough.